### ISSUE TITLE
Ecosystem CI failure: `@__NO_SIDE_EFFECTS__` annotation lost with parenthesized expressions

### ISSUE THREAD

**Issue by camc314** [oxc-project/oxc#17670]
## Description

The `monitor-oxc` ecosystem CI is failing because `@__NO_SIDE_EFFECTS__` annotations are not preserved when the annotated function is wrapped in parentheses.

## CI Failure

From https://github.com/oxc-project/monitor-oxc/actions/runs/20710495854 - the "Test (whitespace)" and "Test (codegen)" jobs fail due to annotation loss.

This pattern appears in Vue's source code:

```javascript
/* @__NO_SIDE_EFFECTS__ */ ((options, extraOptions) => {
  return defineCustomElement(options, extraOptions, hydrate);
})
```

After parsing and codegen, the annotation is stripped:

```javascript
((options, extraOptions) => {
  return defineCustomElement(options, extraOptions, hydrate);
});
```

## Root Cause

Two issues in the parser:

1. **Lexer checkpoint does not preserve trivia flags** - `LexerCheckpoint` does not save/restore `has_pure_comment` and `has_no_side_effects_comment`. When `lookahead()` is used for arrow function detection, it consumes tokens (resetting the flags), and after rewinding the flags remain reset.

2. **`parse_parenthesized_expression` does not propagate the annotation** - When parsing `((x) => x)`, the outer parenthesis handler needs to capture the annotation flag before bumping the `(` token, and apply it to the inner arrow/function expression.

## Possible Solutions

1. **Fix lexer checkpoint** - Save and restore `has_pure_comment` and `has_no_side_effects_comment` in `LexerCheckpoint` so they survive `lookahead()` operations

2. **Fix `parse_parenthesized_expression`** - Capture the annotation flag before bumping `(` and propagate it to nested arrow/function expressions

**Comment by camc314** [comment]
Esbuild: https://esbuild.github.io/try/#dAAwLjI3LjIAAGNvbnN0IGZuID0gLyogQF9fTk9fU0lERV9FRkZFQ1RTX18gKi8gKChvcHRpb25zLCBleHRyYU9wdGlvbnMpID0+IHsKICByZXR1cm4gZGVmaW5lQ3VzdG9tRWxlbWVudChvcHRpb25zLCBleHRyYU9wdGlvbnMsIGh5ZHJhdGUpOwp9KQ

Rollup: https://rollupjs.org/repl/?version=4.55.1&shareable=eyJleGFtcGxlIjpudWxsLCJtb2R1bGVzIjpbeyJjb2RlIjoiXG5leHBvcnQgeyBmbiB9IGZyb20gJy4vcXV4JzsiLCJpc0VudHJ5Ijp0cnVlLCJuYW1lIjoibWFpbi5qcyJ9LHsiY29kZSI6ImV4cG9ydCBjb25zdCBmbiA9IC8qIEBfX05PX1NJREVfRUZGRUNUU19fICovICgob3B0aW9ucywgZXh0cmFPcHRpb25zKSA9PiB7XG4gIHJldHVybiBkZWZpbmVDdXN0b21FbGVtZW50KG9wdGlvbnMsIGV4dHJhT3B0aW9ucywgaHlkcmF0ZSk7XG59KVxuXG5cbmV4cG9ydCBjb25zdCBmbjIgPSAvKiBAX19OT19TSURFX0VGRkVDVFNfXyAqLyAoZm9vKCksIGJhcik7IiwiaXNFbnRyeSI6ZmFsc2UsIm5hbWUiOiJxdXguanMifV0sIm9wdGlvbnMiOnt9fQ==

We have decideed that, to fix this issue, we should propogate these comments to the content inside the parenthesis

We are going to (possibly) follow this up with a change to introduce `warnings` when a `/* @__NO_SIDE_EFFECTS__ */` is in a position that means it could not be attached to a node.

**Comment by sapphi-red** [comment]
#17711
